File: performanceComparisonBioinformatics/C/needleman.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <algorithm>
using namespace std;

const int gap=2;
const int mut=1;
int needleman(string,string);
int minF(int,int);

int main (int argc, char** argv) {
	

	string preFile1="fastaFiles/";
	preFile1+=argv[1];
	const char *f1 = preFile1.c_str();
    
	std::stringstream ss;
	string line;
	ifstream myfile (f1);
	if (myfile.is_open())
	{
		int i=0;
		while ( getline (myfile,line) )
		{
		  if(i==0){
		  		i++;	
			}
			else{
				ss << line ;
			}
		}
		myfile.close();
	}
	else cout << "Unable to open file"; 
	std::string s1 = ss.str();
	//cout<<s1<<"\n";
	
	
	string preFile2="fastaFiles/";
	preFile2+=argv[2];
	const char *f2 = preFile2.c_str();
	
	std::stringstream ss2;
	string line2;
	ifstream myfile2 (f2);
	if (myfile2.is_open())
	{
		int i=0;
		while ( getline (myfile2,line2) )
		{
			if(i==0){
		  		i++;	
			}
			else{
				ss2 << line2 ;
			}
		}
		myfile2.close();
	}
	else cout << "Unable to open file"; 
	
	int m=minF(1,2);
	std::string s2 = ss2.str();
	//cout<<s2<<"\n";
	int score=needleman(s1,s2);
	cout<<score;
	int val=std::min(1,2);
	//cout<<val;
	return score;
}

int needleman(string s1,string s2){
	int len1=s1.length()+1;
	int len2=s2.length()+1;
	int matrizDinamica[len1][len2];
	
	for(int i=0;i<s1.length()+1;i++){
		for(int j=0;j<s2.length()+1;j++){
			matrizDinamica[i][j]=0;
		}
	}	
	
	for(int i=0;i<len1;i++){
		matrizDinamica[i][0]=i*gap;
		
	}
	
	for(int i=0;i<len2;i++){
		matrizDinamica[0][i]=i*gap;
	
	}
	
	for (int i=0;i<len1;i++){
		for (int j=0;j<len2;j++){
			int upVal=matrizDinamica[i-1][j];
			int leftVal=matrizDinamica[i][j-1];
			int diagVal=matrizDinamica[i-1][j-1];
			
			int costUp=upVal+2;
			int costLeft=leftVal+2;
			int costDiag=diagVal;
			if(s1[i-1]!=s2[j-1]){
				costDiag=costDiag+1;
			}
			int min1=std::min(costDiag,costUp);
			int min2=std::min(min1,costLeft);
			if(min2>0){
				matrizDinamica[i][j]=min2;
			}
		}
	}
	return matrizDinamica[len1-1][len2-1];
}

int minF(int a, int b){
	int val=a;
	if(a<b){
		val=a;
	}
	else{
		val=b;
	}
	return val;
}

File: performanceComparisonBioinformatics/Rust/needleman.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::cmp;
use std::env;
const GAP: i32 = 2;
const MUT: i32=1;
fn main() {
    let args: Vec<String> = env::args().collect();
    //println!("{:?}", args);



    let dna1="ATCGATCGATCGATCGATCGAAAAAAAAAAAAA";
    let dna2="CGTACGTACGTACGTAAAAAAAAAAAAAA";
    needelman(dna1,dna2);
    let cad1=readFile("fastaFiles/".to_string()+&args[1].to_owned());
    //println!("{}",cad1);
    let cad2=readFile("fastaFiles/".to_string()+&args[2].to_owned());
    //println!("{}",cad2);
    needelman(&cad1.to_owned(),&cad2.to_owned());
}

fn readFile(file1:String) ->String{
    let mut ans="".to_string();
    let filename = file1;
    // Open the file in read-only mode (ignoring errors).
    let file = File::open(filename).unwrap();
    let reader = BufReader::new(file);

    // Read the file line by line using the lines() iterator from std::io::BufRead.
    for (mut index, line) in reader.lines().enumerate() {
        let line = line.unwrap(); // Ignore errors.
        // Show the line and its number.
        if(index!=0){
        //println!("{}. {}", index + 1, line);
        ans+=&line;
        }
        else{
            index+=1;
        }
    }
    return ans;
}

fn needelman(cadena1:&str,cadena2:&str){
    let len_cad_1=cadena1.chars().count();
    let len_matrix=(len_cad_1)+1;
    //println!("{}",cadena1);
    let len_cad_2=cadena2.chars().count();
    let len_matrix_2=(len_cad_2)+1;
    //println!("{}",cadena2 );
    let mut matriz_dinamica = vec![vec![0i32; len_matrix_2]; len_matrix];
    matriz_dinamica[0][0]=0;
    let mut x = 0; // mut x: i32
    let mut done = false; // mut done: bool
    
    while !done {
        if x == matriz_dinamica.len()-1 {
            done = true;
        }
        matriz_dinamica[x][0]=(x as i32) *GAP  ;
        //println!("{}",matriz_dinamica[x][0]);
        x += 1;
    }
    
    let mut z=0;
    done=false;
    while !done {
        if z == matriz_dinamica[0].len()-1 {
            done = true;
        }
        matriz_dinamica[0][z]=(z as i32) *GAP  ;
        //println!("{}",matriz_dinamica[0][z]);
        z += 1;
    }
    let ch1=cadena1.chars();
    let ch2=cadena2.chars();
    let mut i=1;
    while i<matriz_dinamica.len(){
        let mut j=1;
        while j<matriz_dinamica[i].len(){
            let upVal=matriz_dinamica[i-1][j];
            let leftVal=matriz_dinamica[i][j-1];
            let diagVal=matriz_dinamica[i-1][j-1];

            let costUp=upVal+GAP;
            let costLeft=leftVal+GAP;
            let mut costDiag=diagVal;
            //println!("{}",cadena1.chars().nth(i-1).unwrap()==cadena2.chars().nth(j-1).unwrap());
            if(cadena1.chars().nth(i-1).unwrap()!=cadena2.chars().nth(j-1).unwrap()){
                costDiag+=MUT;
            }
            if(costUp==0){
                //println!("{}",i-1);
            }
            if(costLeft==0){
                //println!("{}",j-1);
            }
            if(costDiag==0){
                //println!("{}",i-1);
                //println!("{}","----");
                //println!("{}",j-1);
            }
            let min1=cmp::min(costUp, costLeft);
            let min=cmp::min(min1,costDiag);
            matriz_dinamica[i][j]=min;
            j+=1;
        }
        i+=1;
    }

    println!("{}",matriz_dinamica[len_cad_1][len_cad_2]);
    
}